blockchain_wasi.md

```diff
@@ -0,0 +1,200 @@
+  `environ_get` call. This includes context of the exeuction in
+  a blockchain context, including the gas constraints, the
+  initiator of execution, and metadata of the transaction
+  initiating the execution. Such metadata SHOULD be exposed as
+  environmental variables.
+
+5. Standards Considerations
+
+5.1 File System Nodes
+
+  Blockchain-specific functionality is exposed in the `/opt`
+  namespace where it does not conflict with existing file
+  system semantics.
+
+  The home directory MUST be writable, and is mapped to persistant
+  storage that will be committed with a transaction and
+  subsequently available to future transactions.
+
+  The `/tmp` directory SHOULD be used to provide a namespace that is
+  writable but ephemeral, and will not be saved between
+  transactions.
+
+  The remainder of the file system presented SHOULD be read-only,
+  and any attempts at writes SHOULD fail.
+
+  The virtual file system translates paths to addresses in
+  the underlying key-value store. In WASI, the paths and byte
+  offsets are supplied to the Wasm runtime via `__wasi_fd_open`,
+  `__wasi_fd_seek`. The runtime associates the key with the file
+  descriptor so that it can service future reads and writes
+  received from `__wasi_fd_(read|write)`.
+
+5.1.1 Deviations from Expected File System Behavior
+
+  Directory listing of the `/opt/CHAIN` directory will list 0
+  items, since the full namespace cannot meaningfuly be represented.
+  Attempts to `fd_fdstat_set_rights` (change permissions), or
+  `_filestat_set_times` (changing metadata), on blockchain-namespace
+  files will be a failure that aborts execution for safety.
+
+5.2 Socket Address and Interoperability
+
+  Opening of a socket to another blockchain executable via a
+  FD obtained from `/opt/CHAIN/ADDRESS/sock` allows for cross
+  contract calls. Data sent over the socket via `sock_send` will
+  be used as the input available on the `stdin` descriptor available
+  to the other execution.
+  The first `sock_recv` call will begin execution and block until
+  output is available.
+
+5.3 Environment
+
+  We expose the environmental variables
+  * address
+  * gas_left
+  * sender
+  * value
+  * balance
+  * block_number
+  * $HOME is set to `/home/ADDRESS`, the current contract address.
+
+5.4 Process Signals
+
+  Any signal issued via `proc_raise`, or `proc_exit` will end the process.
+  A call that is not a 0 exit status to `proc_exit` will abort and revert
+  the execution.
+
+5.5 Logs & Events
+
+  Logs in a blockchain context allow a contract to expose a structured
+  message to the outside world. Since these messages are structured,
+  they do not map cleanly to file system items. Instead, they SHOULD be
+  considered as a chain-defined protocol, similar to IPC messages
+  a program would pass to a system logging service. An socket for
+  communication to a logging service to ultimately emit externally
+  visible events SHOULD live at `/srv/CHAIN/log.sock` if such a service
```

Comment:
Should this be `/opt/` rather than `/srv/`?

Reply:
```suggestion
  visible events SHOULD live at `/opt/CHAIN/log.sock` if such a service
```